src/expr/core/src/window_function/state/buffer.rs

```diff
@@ -258,19 +171,171 @@ impl<K: Ord, V: Clone> WindowBuffer<K, V> {
             .drain(0..min_needed_idx)
             .map(|Entry { key, value }| (key, value))
     }
+
+    fn maintain_delta(&mut self, old_outer: Range<usize>, new_outer: Range<usize>) {
+        debug_assert!(self.frame_exclusion.is_no_others());
+
+        let (outer_removed, outer_added) = range_diff(old_outer.clone(), new_outer.clone());
+        let delta = self.curr_delta.as_mut().unwrap();
+        for idx in outer_removed.iter().cloned().flatten() {
+            delta.push((Op::Delete, self.buffer[idx].value.clone()));
+        }
+        for idx in outer_added.iter().cloned().flatten() {
+            delta.push((Op::Insert, self.buffer[idx].value.clone()));
+        }
+    }
+
+    fn recalculate_left_right(&mut self) {
+        let buffer_ref = BufferRefMut {
+            buffer: &self.buffer,
+            curr_idx: &mut self.curr_idx,
+            left_idx: &mut self.left_idx,
+            right_excl_idx: &mut self.right_excl_idx,
+        };
+        self.window_impl.recalculate_left_right(buffer_ref);
+    }
+}
+
+#[derive(Educe)]
+#[educe(Clone, Copy)]
+pub(super) struct BufferRef<'a, K: Ord, V: Clone> {
+    buffer: &'a VecDeque<Entry<K, V>>,
+    curr_idx: usize,
+    left_idx: usize,
+    right_excl_idx: usize,
+}
+
+pub(super) struct BufferRefMut<'a, K: Ord, V: Clone> {
+    buffer: &'a VecDeque<Entry<K, V>>,
+    curr_idx: &'a mut usize,
+    left_idx: &'a mut usize,
+    right_excl_idx: &'a mut usize,
+}
+
+pub(super) trait WindowImpl {
```

Comment:
Can you please add some comments for this trait and its member functions? I feel difficult to understand 🥹

Reply:
Added